### ISSUE TITLE
Rendering issues when loading Tangram offscreen via Mapzen.js

### ISSUE THREAD

**Issue by meetar** [mapzen/mapzen.js#322]
Reported from @mrmetlhed via https://twitter.com/mrmetlhed/status/809123614943576064 :

Render errors are seen if a map is loaded offscreen:

![](https://pbs.twimg.com/media/CzqVkuEUoAANTbi.jpg)

Gets in a sad state and refresh is required:

![](https://pbs.twimg.com/media/CzqWLACUoAA0boM.jpg)

But if a delay is added to the layer.addto(), it works:

![](https://pbs.twimg.com/media/CzqXmDuUAAA4z6M.jpg)

I will attempt to replicate the behavior and post a link to a test scene here.

**Comment by hanbyul-here** [comment]
Do you have clue that what part of Tangram is not happy with offscreen loading based on warnings?

**Comment by meetar** [comment]
Not yet! I'll know more after I have a test case.

**Comment by MrMetlHed** [comment]
Here's something that 'works' 50% of the time: https://gist.github.com/MrMetlHed/0142ff151e32ce1b1935fb578c24603f

Most likely to work if you clear your cache, refresh, and then scroll (at least for me.) In this case if I uncomment the listener for tangramloaded the glitch disappears, but previously for me it didn't matter. The code was a little more complex in that case, but I can't figure out how to make it happen again.

Hope that helps.

**Comment by meetar** [comment]
Confirmed – investigating now.

**Comment by hanbyul-here** [comment]
Thanks for the gist @MrMetlHed  ! 

I noticed you are setting up the map with Refill style first, and replace Refill with `data.yaml`  after `tangramloaded`event is fired. Can I ask what you tried to do there?  

mapzen.js tries to create `TangramLeafletLayer` and add it to the map when Tangram is loaded, (when `tangramloaded`is fired), which means the timing `Refill` layer is created & added and timing `data.yaml`is created & added to the map can be so close to each other. We suspect this is the reason of the glitch.

**Comment by MrMetlHed** [comment]
I have to confess that we didn't really know the best way to use Tangram at the time I ran into this problem.

The idea was basically to load the base map (the Refill style) and then drop our data on top of it with some transparency (not in this version of `data.yaml`, but was updated to have transparent colors at one point) so people could see the underlying map and figure out exactly the area the data affected. This is sort of how we'd use something like Mapbox or Google Maps in the past, so we were going with what we knew.

Eventually when we figured out that we should just merge our YAML with the base style we wanted (we wanted to re-order Refill slightly so our layer was between different things than it would have been on the default), we could avoid the issue entirely by loading it all at once using a straight Leaflet map with a single Tangram layer. Which is what our final version looked like.

Hope that answers your question.

**Comment by meetar** [comment]
I think we're also running into an architectural issue, that being only one Tangram map is currently allowed per "browsing context" – you can read more about it here: https://github.com/tangrams/tangram-docs/blob/gh-pages/pages/Tangram-Overview.md#multiple-maps

The idea is that to use our basemaps and then style things on top of them, for now we recommend using `import` to import the basemap you want to use into your scene file, eg:

`import: https://mapzen.com/carto/refill-style-no-labels/refill-style-no-labels.yaml`

Then you can order your own data on top of that.

**Comment by meetar** [comment]
Looks like you reached that same conclusion :) We're planning more examples and tutorials around using our basemaps – sorry for the confusion and thanks very much for the feedback!

**Comment by MrMetlHed** [comment]
Yeah, no problem. More tutorials would probably help -- we eventually found what would work best for us in blog posts, but the documentation is sometimes tricky to follow if you don't know what you're looking for.

I think basic users may go about trying to do exactly what I did and have trouble. Importing other styles isn't common in other libraries. If there was a way to warn users in the console if they try to do this that'd be pretty helpful.

All told, merging the styles is actually better for us in the long run -- I can see us making an entire style of our own off of the vector tiles and merging our data at appropriate layers. It's quite powerful, just very different from how we'd normally go about building something like this.

Thanks for all your help and responsiveness. It's really a great library.
